code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -56,12 +90,18 @@ public void pingReply(Ping ping) {
         }
 
         long agentId = Long.parseLong(ping.getResourceId());
-        Map<String, String> knownInstances = instanceCache.getUnchecked(agentId);
+        Map<String, Instance> knownInstances = instanceCache.getUnchecked(agentId);
+
+        Agent agent = objectManager.loadResource(Agent.class, agentId);
+        Map<String, Host> hosts = agentDao.getHosts(agent.getId());
```

Comment:
I would deal with the situation when an agent has no hosts.  Also cache the response so that we don't query for each ping